subprojects/core/src/main/java/org/gradle/internal/fingerprint/classpath/impl/ClasspathCompareStrategy.java

```diff
@@ -41,56 +43,129 @@ private ClasspathCompareStrategy() {
 
     @Override
     protected boolean doVisitChangesSince(ChangeVisitor visitor, Map<String, FileSystemLocationFingerprint> currentSnapshots, Map<String, FileSystemLocationFingerprint> previousSnapshots, String propertyTitle, boolean includeAdded) {
-        Iterator<Map.Entry<String, FileSystemLocationFingerprint>> currentEntries = currentSnapshots.entrySet().iterator();
-        Iterator<Map.Entry<String, FileSystemLocationFingerprint>> previousEntries = previousSnapshots.entrySet().iterator();
-        while (true) {
-            if (currentEntries.hasNext()) {
-                Map.Entry<String, FileSystemLocationFingerprint> current = currentEntries.next();
+        TrackingVisitor trackingVisitor = new TrackingVisitor(visitor);
+        ChangeState changeState = new ChangeState(propertyTitle, includeAdded, trackingVisitor, currentSnapshots, previousSnapshots);
+
+        while (trackingVisitor.isConsumeMore() && changeState.hasMoreToProcess()) {
+            changeState.processChange();
+        }
+        return trackingVisitor.isConsumeMore();
+    }
+
+    private static class TrackingVisitor implements ChangeVisitor {
+        private final ChangeVisitor visitor;
+        private boolean consumeMore = true;
+
+        private TrackingVisitor(ChangeVisitor visitor) {
+            this.visitor = visitor;
+        }
+
+        @Override
+        public boolean visitChange(Change change) {
+            if (consumeMore) {
+                consumeMore = visitor.visitChange(change);
+            }
+            return consumeMore;
+        }
+
+        public boolean isConsumeMore() {
+            return consumeMore;
+        }
+    }
+
+    private static class ChangeState {
```

Comment:
I think the class makes things better since I can extract methods for modified/removed/added and the methods are much clearer to me than requiring to read the code we had before.